### ISSUE TITLE
For consideration: stashing the loaded `refresh_token` object onto the `request` instance

### ISSUE THREAD

**Issue by jvanasco** [oauthlib/oauthlib#592]
This discussion is forked from #557

When issuing a refresh token, there are two related potential improvements of the existing code and/or implementation guidelines.

1. If the request for refreshing a token is successful,  the `save_bearer_token` function does not natively have access to the object correlated to the `refresh_token` object. At the very least, this means that `save_bearer_token` does not easily know what kind of authorization grant was associated with the original token - which could have been issued to a user of the Client via an `authorization_code` or issued to the owner of the Client via `client_credentials`.  The `refresh_token` unique identifier string is available on the `request` instance, but it will require either (i) a subsequent database query, or (ii) a developer previously stashing the refresh token object onto the `request` in a private namespace.

2. The existing code for refreshing a token is "less than optimal" for most implementations, because it leads to repeated database queries.  The existing code is extremely flexible and well suited for extremely large scale deployments because it allows for fine-grained data selection -- something that must still be supported -- but popular oauthlib deployments don't always optimize this. For example, the following validator functions all require loading a refresh token or parts of the refresh token's record: 

* `validate_refresh_token`
* `get_original_scopes`
* `is_within_original_scope`

Additionally, consider this code block in `grant_types.refresh_token.RefreshTokenGrant.validate_token_request` 
 ( https://github.com/oauthlib/oauthlib/blob/master/oauthlib/oauth2/rfc6749/grant_types/refresh_token.py#L115-L137 )

Notice how each of the above functions is executed in order:

    if not self.request_validator.validate_refresh_token(
    ....
    self.request_validator.get_original_scopes(
    ...
    self.request_validator.is_within_original_scope(

A naive implementation of a library like `flask-oauthlib` would end up with multiple database queries because of this design.  Unless the refresh-token is a primary key (and it is likely not), most ORMs will not be able to tell they have previously loaded this record and a new database call will be issued.

I've been thinking a lot about this, and honestly can't think of good solutions that solve all the needs.  I'll list a bunch of my thoughts below:

* does it make sense to augment the dict in `save_bearer_token` with the type of of token it is (e.g. authorization_code or client_credentials, etc) AND it's lineage (if a refresh token)?  Much of this information is available by inspecting various request attributes, but it's not easily accessible and it is too easy to deploy something with oauthlib and then realize "crap, i actually needed that info"

* does it make sense to create a dedicated request attribute for library users to use as a workspace? 
 this is essentially the solution suggested in #557, but instead of tacking things on to existing objects my suggestion is to reserve a namespace like `request.implementer`.

* does it make sense to stash some sort of `request.legacy_token` object that would be an official/expected cache of the refresh_token instance?  This is admittedly weird, because the current codebase does not have a concept of a `refresh_token` object - it is limited to just validating the token string and loading (or checking) the scopes associated with it.  Sites with high traffic or massive amounts of permissions obviously benefit from the current design; however signs with lower traffic/permissions that use generic packages out-of-the-box are likely taking an extra bit of database queries.

-----

To recap, the two concepts are:

* enable `save_bearer_token`, on a refresh, to be natively aware of the previous Bearer token's grant type
* make it easier/more-well-known for users to lower their database queries with regards to refresh tokens.
